src/engine/config/shop-config.ts

```diff
@@ -52,58 +52,99 @@ export class Shop {
         const itemStock = this.container.amount(item.gameId);
 
-        if(itemSoldHere) {
-            const foundStock = this.originalStock.find(stock => stock && stock[0] === itemKey);
-
+        if (itemSoldHere) {
+            const foundStock = this.originalStock.find(stock => stock && stock.itemKey === itemKey);
             if (foundStock) {
-                originalStockAmount = foundStock[1];
+                originalStockAmount = foundStock.amount;
             }
-        } else if(!this.generalStore) {
+        } else {
+
+            return -1; // Cannot buy from this shop
+        }
+
+        let finalAmount: number;
+        if (itemStock === originalStockAmount) {
+            finalAmount = Math.round(item.value * this.sellRate);
+        } else if (itemStock > originalStockAmount) {
+            const overstockAmount = (itemStock - originalStockAmount);
+            let shopSellRate = this.sellRate;
+            shopSellRate -= (this.rateModifier * overstockAmount);
+            finalAmount = item.value * shopSellRate;
+
+            finalAmount = Math.round(finalAmount);
+        } else {
+            const understockAmount = (originalStockAmount - itemStock);
+            let shopSellRate = this.sellRate;
+            shopSellRate += (this.rateModifier * understockAmount);
+            finalAmount = item.value * shopSellRate;
+
+            finalAmount = Math.round(finalAmount);
+        }
+
+        const min = item.minimumValue || 1;
+        return finalAmount < min ? min : finalAmount;
+    }
+
+    /**
+     * Price which shop pays player for.
+     * @param item
+     */
+    public getSellToShopPrice(item: ItemDetails): number {
+        const itemKey = item.key;
+        const itemSoldHere: boolean = this.isItemSoldHere(itemKey);
+        let originalStockAmount: number = 0;
+        const itemStock = this.container.amount(item.gameId);
+
+
+        if (itemSoldHere) {
+            originalStockAmount = this.originalStock.find(stock => stock && stock.itemKey === itemKey)?.amount || 0;
+        } else if (!this.generalStore) {
             return -1; // Can not sell this item to this shop (shop is not a general store!)
         }
 
         let finalAmount: number;
 
-        if(itemStock > originalStockAmount) {
+        if (itemStock === originalStockAmount) {
+            finalAmount = Math.round(item.value * this.buyRate);
+        } else if (itemStock > originalStockAmount) {
             const overstockAmount = (itemStock - originalStockAmount);
-
-            if(this.generalStore) {
-                const decrementAmount = 0.075 * overstockAmount;
-                finalAmount = item.lowAlchValue - decrementAmount;
-            } else {
-                let shopBuyRate = this.buyRate;
-                shopBuyRate -= (this.rateModifier * overstockAmount);
-                finalAmount = item.value * shopBuyRate;
-            }
+            let shopBuyRate = this.buyRate;
```

Comment:
Long-term it would be nice to get some of these `buyRate` logic into utility functions, same for `sellRate` and `getBuyFromShopPrice`

I don't think it's a change that needs to be done in this PR though